scipy/special/_cosine_distr.pxd

```diff
@@ -0,0 +1,244 @@
+#
+#     import mpmath
+#
+#     def f(x):
+#         return (mpmath.pi + x + mpmath.sin(x)) / (2*mpmath.pi)
+#
+#     # Note: 40 digits might be overkill; a few more digits than the default
+#     # might be sufficient.
+#     mpmath.mp.dps = 40
+#     ts = mpmath.taylor(f, -mpmath.pi, 20)
+#     p, q = mpmath.pade(ts, 9, 10)
+#
+# The following are the values after converting to 64 bit floating point:
+# p = [0.0,
+#      0.0,
+#      0.0,
+#      0.026525823848649224,
+#      0.0,
+#      -0.0007883197097740538,
+#      0.0,
+#      1.0235408442872927e-05,
+#      0.0,
+#      -3.8360369451359084e-08]
+# q = [1.0,
+#      0.0,
+#      0.020281047093125535,
+#      0.0,
+#      0.00020944197182753272,
+#      0.0,
+#      1.4162345851873058e-06,
+#      0.0,
+#      6.498171564823105e-09,
+#      0.0,
+#      1.6955280904096042e-11]
+#
+
+cdef inline double cosine_cdf_pade_approx_at_neg_pi(double x) nogil:
+    """
+    Compute the CDF of the standard cosine distribution for x close to but
+    not less than -π.  A Pade approximant is used to avoid the loss of
+    precision that occurs in the formula 1/2 + (x + sin(x))/(2*pi) when
+    x is near -π.
+    """
+    cdef double h, h2, h3
+    cdef double numer, denom
+
+    # M_PI is not exactly π.  In fact, float64(π - M_PI) is
+    # 1.2246467991473532e-16.  h is supposed to be x + π, so to compute
+    # h accurately, we write the calculation as:
+    h = (x + M_PI) + 1.2246467991473532e-16
+    h2 = h*h
+    h3 = h2*h
+    numer = (h3*(0.026525823848649224
+                 + h2*(-0.0007883197097740538
+                       + h2*(1.0235408442872927e-05 
+                             + h2*(-3.8360369451359084e-08)))))
+    denom = (1.0 + h2*(0.020281047093125535
+                       + h2*(0.00020944197182753272
+                             + h2*(1.4162345851873058e-06
+                                   + h2*(6.498171564823105e-09
+                                         + h2*(1.6955280904096042e-11))))))
+    return numer / denom
+
+
+cdef inline double cosine_cdf(double x) nogil:
+    """
+    cosine distribution cumulative distribution function (CDF).
+    """
+    if x >= M_PI:
+        return 1
+    elif x < -M_PI:
+        return 0
+    elif x < -1.6:
+        return cosine_cdf_pade_approx_at_neg_pi(x)
+    else:
+        return 0.5 + (x + sin(x))/(2*M_PI)
+
+
+# The CDF of the cosine distribution is
+#     p = (pi + x + sin(x)) / (2*pi),
+# We want the inverse of this.
+#
+# Move the factor 2*pi and the constant pi to express this as
+#     pi*(2*p - 1) = x + sin(x)
+# Then if f(x) = x + sin(x), and g(x) is the inverse of f, we have
+#     x = g(pi*(2*p - 1)).
+
+# The coefficients in the functions _p2 and _q2 are the coefficients in the
+# Pade approximation at p=0.5 to the inverse of x + sin(x).
+# The following steps were used to derive these:
+# 1. Find the coefficients of the Taylor polynomial of x + sin(x) at x = 0.
```

Comment:
Note that `special._precompute` has a canned method for reverting series: https://github.com/scipy/scipy/blob/master/scipy/special/_precompute/utils.py#L20